### PR TITLE
ArC - introduce the CurrentContextFactory abstraction

### PR DESCRIPTION
**Pull request by mkouba**
- this API allows quarkus to supply an optimized backend for any
non-shared context of a normal scope
- currently it's only used by the request context
- an extension can provide a custom factory via
ContextReferenceFactoryBuildItem

### CODE REVIEW

**Comment by mkouba** [comment]
The workflow is simple: the request context will use a [ContextReference](https://github.com/quarkusio/quarkus/pull/22319/files#diff-32b0d6704be80e03145cb21a80d3ad58fe3bb2ef00c9e7f49290f2106485dcdd) created via the [ContextReferenceFactory](https://github.com/quarkusio/quarkus/pull/22319/files#diff-70efb64833fcd99b4704fea01d8714d5dfcc720976ff3905b24be21c7abae24a). The default impl is using `TheadLocal` variables. A quarkus extension, e.g. vertx, can provide an optimized implementation of `ContextReference` via `ContextReferenceFactoryBuildItem`.

**Comment by mkouba** [comment]
Hm, maybe `ContextStateReference` or `StateReference` would be a better name...

**Comment by stuartwdouglas** [comment]
I think this looks good. I will look at building on top of this.

**Comment by geoand** [comment]
👍

**Comment by manovotn** [comment]
+1, this looks great.

> Hm, maybe ContextStateReference or StateReference would be a better name...

I'd prefer `ContextStateReference` as well just to make it super clear but it's nitpicking.

One more note - the `ThreadLocalContextReference` and the factory for it could IMO be directly in Arc. Especially if this SPI can be used by any custom normal scope whatsoever.

**Comment by mkouba** [comment]
> One more note - the `ThreadLocalContextReference` and the factory for it could IMO be directly in Arc.

It is in ArC... Do you mean the `io.quarkus.arc` API package? I don't think it's necessary because a custom context can just obtain the factory via `Arc.container().getContextReferenceFactory()` and the default impl is threadlocal-based.

**Comment by manovotn** [comment]
Yes, I meant the API package but failed to mention it, sorry.

>  don't think it's necessary because a custom context can just obtain the factory via Arc.container().getContextReferenceFactory()

Ok, fair enough, I didn't realize that. Consider my complaint obsolete :-)

**Comment by FroMage** [comment]
SR-CP had this already created in a new Maven project with zero dependency:

The idea was that you declared your storage need with https://github.com/smallrye/smallrye-context-propagation/blob/main/storage/src/main/java/io/smallrye/context/storage/spi/StorageDeclaration.java which allowed our extension to pick up all storage slots: https://github.com/FroMage/quarkus/commit/b90e4e51ce75f23582ffd1feed808b57167910eb

Here's how ArC could then declare its storage requirements and use the new storage: https://github.com/FroMage/quarkus/commit/372c2295844e0dd890dd506d9a7f21fd97428f28

Note that we went with `ThreadLocal` because that minimises other libraries changes to replacing `new ThreadLocal<>()` with `StorageManager.threadLocal(RequestContextStorageDeclaration.class);`

**Comment by FroMage** [comment]
On non-Quarkus, the default storage manager hands out real `ThreadLocal` instances: https://github.com/smallrye/smallrye-context-propagation/blob/main/storage/src/main/java/io/smallrye/context/storage/impl/DefaultStorageManager.java

But for Quarkus, the ThreadLocals will store stuff in custom thread fields: https://github.com/FroMage/quarkus/commit/b90e4e51ce75f23582ffd1feed808b57167910eb#diff-f36606de9fadce013fb441b209205c1ff4b55eaae93d2a1d5c3fa7549e6003ecR100 . We only have to change that to store them in the Vert.x context.

**Comment by mkouba** [comment]
This abstraction is an ArC-specific bit that is only used by ArC contexts. I don't want to add the `smallrye-context-propagation-storage` dependency in the ArC core, i.e. `io.quarkus.arc:arc`. Ofc, we can implement this in quarkus-arc.

**Comment by FroMage** [comment]
Well, how do you get other projects to use the ArC storage then if there's no module they can import to get it?

**Comment by mkouba** [comment]
> Well, how do you get other projects to use the ArC storage then if there's no module they can import to get it?

I'm not sure I understand. Other projects can simply use the beans stored in the CDI contexts. We don't want them to use the ArC internals...

**Comment by stuartwdouglas** [comment]
Here is something (a bit hacked up) built on top of this that should be suitable for testing:
https://github.com/quarkusio/quarkus/compare/main...stuartwdouglas:arc-context-store?expand=1

The interesting bit here is that to get the result we want we need to effectively deliberately 'leak' the context by ignoring the 'remove' method. As we know the context is tied to the request this should be safe, as the data will now have the same lifecycle as the context data.

@Sanne any chance you could benchmark this?

**Comment by FroMage** [comment]
> I'm not sure I understand. Other projects can simply use the beans stored in the CDI contexts. We don't want them to use the ArC internals...

That's precisely my point: we have an independent module already with zero dependencies, that abstracts `ThreadLocal`s, that JTA and RESTEasy and any lib can use without depending on CDI or ArC, and even ArC can use, and then we can store all these in the Vert.x context and call it a day.

**Comment by mkouba** [comment]
> That's precisely my point: we have an independent module already with zero dependencies, that abstracts `ThreadLocal`s, that JTA and RESTEasy and any lib can use without depending on CDI or ArC, and even ArC can use, and then we can store all these in the Vert.x context and call it a day.

Well, this PR does not prevent this solution except that I want to have the ArC core more pluggable and not tied to the API of the `smallrye-context-propagation-storage`. We could add the `smallrye-context-propagation-storage` dependency to the `quarkus-arc` and implement `StorageDeclaration` there...

**Comment by FroMage** [comment]
But `ContextReference` and `smallrye-context-propagation-storage` appear to do the exact same thing. Perhaps you're thinking that because of its name that module depends on CP, but it doesn't.

**Comment by mkouba** [comment]
> But ContextReference and smallrye-context-propagation-storage appear to do the exact same thing.

Almost. A `ContextReference` impl does not need to use threadlocals at all...

**Comment by manovotn** [comment]
@mkouba @Sanne @stuartwdouglas I have rebased this branch and resolved conflicts with the recent changes to req. context propagation. I think we should restart the discussion here as this is still a valuable addition.

Some time ago, Stuart created a testable PoC for this PR, see https://github.com/quarkusio/quarkus/pull/22319#issuecomment-998356103

We can still take that commit and slap it on top of this rebased branch just fine.
@Sanne would you be able to benchmark it to see the actual numbers this would give us?

**Comment by mkouba** [comment]
I've rebased this PR one more time and also renamed the `ContextReference` abstraction to `CurrentContext` (which is IMO closer to the spec wording).

Unfortunately, this means that Stuart's PoC would have to be rebased as well.

In any case, I believe that this abstraction makes a lot of sense and I'd like to have it merged no matter whether we'll have some benchmarks available or not (and given the current workload it seems that we won't have them any time soon ;-).

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java [hunk 1]
```diff
@@ -95,7 +97,9 @@
 
     private volatile ExecutorService executorService;
 
-    public ArcContainerImpl() {
+    private final CurrentContextFactory contextReferenceFactory;
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java:100]:
```suggestion
    private final CurrentContextFactory currentContextFactory;
```

?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ThreadLocalContextReference.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import io.quarkus.arc.CurrentContext;
+import io.quarkus.arc.InjectableContext.ContextState;
+
+/**
+ * {@link ThreadLocal} implementation of {@link CurrentContext}.
+ *
+ * @param <T>
+ */
+final class ThreadLocalContextReference<T extends ContextState> implements CurrentContext<T> {
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ThreadLocalContextReference.java:11]:
```suggestion
final class ThreadLocalCurrentContext<T extends ContextState> implements CurrentContext<T> {
```

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ThreadLocalContextReferenceFactory.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import io.quarkus.arc.CurrentContextFactory;
+import io.quarkus.arc.InjectableContext.ContextState;
+import java.lang.annotation.Annotation;
+
+/**
+ * The default implementation makes use of {@link ThreadLocal} variables.
+ *
+ * @see ThreadLocalContextReference
+ */
+final class ThreadLocalContextReferenceFactory implements CurrentContextFactory {
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ThreadLocalContextReferenceFactory.java:13]:
```suggestion
final class ThreadLocalCurrentContextFactory implements CurrentContextFactory {
```

**Review by Ladicek — APPROVED** [review]
Just a few suggestions to unify naming, otherwise LGTM!

**Comment by mkouba** [comment]
> Just a few suggestions to unify naming, otherwise LGTM!

Thanks. Comments are addressed.

**Review by manovotn — APPROVED** [review]

**Pull request merged**